runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1825,6 +1832,106 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
                return;
                }
             }
+         else if (isNonNullJavaLangClass)
+            {
+            if (!performTransformation(comp(), "%sTransforming %s on node %p to load component type inline\n", OPT_DETAILS, signature, node))
+               break;
+
+            // Consider two cases:
+            //
+            // (i)  The operand is an instance of java.lang.Class indirectly loaded through a vft.  In that case,
+            //      we can work with the vft directly
+            // (ii) The operand is definitely a non-null instance of java.lang.Class.  In that case, load the
+            //      J9Class from the java.lang.Class by way of <classFromJavaLangClass>
+            //
+
+            TR::SymbolReference *symRef = classChild->getOpCode().hasSymbolReference() ? classChild->getSymbolReference() : NULL;
+            TR::SymbolReference *jlcFromClassSymRef = comp()->getSymRefTab()->findOrCreateJavaLangClassFromClassSymbolRef();
+            TR::Node *classOperand = NULL;
+
+            if ((symRef != NULL) && (symRef == jlcFromClassSymRef))
+               {
+               classOperand = classChild->getFirstChild();
+               }
+            else
+               {
+               classOperand = TR::Node::createWithSymRef(TR::aloadi, 1, 1, classChild,
+                                 comp()->getSymRefTab()->findOrCreateClassFromJavaLangClassSymbolRef());
+               }
+
+            TR::Node *loadComponentTypeNode = comp()->fej9()->loadArrayClassComponentType(classOperand);
+            TR::Node *jlcOfComponentTypeNode = NULL;
+
+            if ((classChildConstraint == NULL)
+                || (classChildConstraint->getClassType() == NULL)
+                || !comp()->fej9()->isClassArray(classChildConstraint->getClass()))
+               {
+               // Generate IL of the following form, inlining Class.getComponentType()
+               // without control flow:
+               //
+               //    n99n aselect
+               //    n98n   icmpne
+               //    n97n     iand
+               //    n96n       l2i
+               //    n95n         lloadi <classDepthAndFlags>
+               //    n94n           aloadi <vft> (or aloadi <classFromJavaLangClass>)
+               //    n93n       iconst 0x10000
+               //    n92n     iconst 0
+               //    n91n   aloadi <javaLangClassFromClass>
+               //    n90n     aselect
+               //    n98n       ==> icmpne
+               //    n89n     aloadi <componentClass>
+               //    n94n       ==> aloadi <vft> (or aloadi <classFromJavaLangClass>)
+               //    n94n     ==> aloadi <vft> (or aloadi <classFromJavaLangClass>)
+               //    n88n   aconst NULL
+               //
```

Comment:
@vijaysun-omr, I saw a comment that you had added here that disappeared as I was typing up a response.  I thought I would still enter the response as additional comments I might add to the code.

The last appearance of `n94n` is a sort of placeholder for the non-array case.  I wanted to produce IL that would not require control flow that would choose (the `aselect` at `n99n`) between the result of loading `<javaLangClassFromClass>` for the `<componentClass>` field of a `j9ArrayClass` and NULL.  But because all three operands of an `aselect` are evaluated unconditionally, I needed the `<javaLangClassFromClass>` to have a valid `j9Class` it could operate on for both the array case and the non-array case.  So the `aselect` at `n90n` yields either the contents of the `<componentClass>` field or the original `j9Class` from `n94n`, just so the `aloadi <javaLangClassFromClass>` would always succeed.  When it uses the `j9Class` from `n94n` to perform the `aloadi <javaLangClassFromClass>`, the result is ultimately discarded.  So it's potentially wasteful in the case that the operand is not an array class, under the assumption that `Class.getComponentType()` is unlikely to be called for non-array classes.

Maybe it would be easier to understand if I had also written it out this way, as pseudo-code?

```
j9Class origClassn94n = ...;
bool isArrayClassn98n = (((int) origClass->classDepthAndFlags) & 0x10000) != 0;
j9Class choice1n89n = ((j9ArrayClass) origClass)->componentClass;
j9Class choice2n94n = origClass;
j9Class aselectn90n = isArrayClassn98n ? choice1n89n : choice2n94n;
java.lang.Class jlcn91n = origClass->javaLangClassFromClass;
java.lang.Class resultn99n = isArrayClassn98n ? jlcn91n : NULL;
```

Other suggestions will be gladly entertained!

Reply:
Thanks Henry. I had asked my prior question when I was looking at the IL trees you had drawn. I subsequently deleted my question since you had explained the reasoning about what I was essentially asking (i.e. wanting a valid class in both the array and non-array case, even though it is only used in one case).